File: Solution/0019.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        recordNodeBack = head
        recordNodeFront = head
        recordCnt = 0
        
        while(recordNodeFront.next != None):
            recordNodeFront = recordNodeFront.next
            recordCnt += 1
            if(recordCnt > n):
                recordNodeBack = recordNodeBack.next
            
        if(recordCnt == 0): # Only 1 node
            return None
        elif(recordCnt < n):
            head = head.next
        else:
            if(recordNodeBack.next == None): # Deleted node is in the tail
                recordNodeBack = recordNodeBack.next
            else:
                recordNodeBack.next = recordNodeBack.next.next
        
        return head
            
            

File: Solution/0019.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dummy = ListNode(-1, head);
        ListNode *tmpNode = &dummy;

        int count = 1;
        while(head != nullptr) {
            if (count > n) tmpNode = tmpNode -> next;
            
            count += 1;
            head = head -> next;
        }

        tmpNode -> next = tmpNode -> next -> next;
        return dummy.next;
    }
};